src/main/java/org/opensearch/security/identity/SecuritySubject.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.security.identity;
+
+import java.security.Principal;
+
+import org.opensearch.common.util.concurrent.ThreadContext;
+import org.opensearch.identity.NamedPrincipal;
+import org.opensearch.identity.Subject;
+import org.opensearch.identity.tokens.AuthToken;
+import org.opensearch.security.support.ConfigConstants;
+import org.opensearch.security.user.User;
+import org.opensearch.threadpool.ThreadPool;
+
+public class SecuritySubject implements Subject {
+
+    private ThreadContext threadContext;
+
+    public SecuritySubject() { }
+
+    public void setThreadContext(ThreadContext threadContext) {
+        this.threadContext = threadContext;
+    }
+    @Override
+    public Principal getPrincipal() {
+        if (threadContext == null) {
+            return NamedPrincipal.UNAUTHENTICATED;
+        }
+        final User user = (User) threadContext.getTransient(ConfigConstants.OPENDISTRO_SECURITY_USER);
+        if (user == null) {
+            return NamedPrincipal.UNAUTHENTICATED;
```

Comment:
When there is no user in the thread context isn't this when we are running as 'root' in OpenSearch?  I really like that we'd be patching this hole, but want to double check there aren't implications

Reply:
That's partially true, what really defines superuser is:

1. Lack or User in the threadcontext
2. Lack or Origin in the threadcontext or Origin === LOCAL
3. No injected roles

See https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/filter/SecurityFilter.java#L258-L266 - this is the block that skips PrivilegeEvaluation after a plugin stashes its threadcontext and executes a TransportAction

I think you are correct though that if a plugin calls on `getSubject()` and there is no User in the threadcontext that that would mean that the plugin is in superuser mode.

Reply:
While I don't like the behavior of falling back to superuser on a null, I think we should keep the relationship subject <-> security user consistent otherwise we might have logging that says the user is unauthenticated alongside system-only changes.

Note; I think changing the default behavior would be very good for the security posture, but we would likely need to make this a breaking change ala 3.0.

Reply:
What about `None` since that's technically the case? If the threadcontext is stashed then there is no subject. Contrast that to doing clientcert auth with a certificate that matches an `admin_dn` configured in `opensearch.yml`. That would truly be the root user.

Reply:
While tangentially related there are a lot of implications of these values that can create confusion, if you'd prefer not to address this now, how about a follow up issue?